### DISCUSSION TITLE
`superclass mismatch for class DashboardController` in development

### DISCUSSION THREAD

**Discussion by vfonic** [activeadmin/activeadmin#8606] [Ideas]
In new Rails 8 app, ActiveAdmin 4 (beta 15 or `master`), after raising an error such as `NoMethodError` in some AA resource class, triggers `superclass mismatch for class DashboardController`. The only mitigation is to restart the server.
The error needs to be raised at the "root" of the class/resource, such as:

```rb
ActiveAdmin.register Post do
  raise 'asd'
end
```

Here are the steps to repro:

```sh
rails new superclass_admin 
cd superclass_admin
bundle install
rails db:prepare
rails g scaffold post title:string
bundle add activeadmin --version "4.0.0.beta15"
rails g active_admin:install --skip-users
rails db:migrate
rails g active_admin:resource Post
rails s
```

After this, open `app/admin/posts.rb` and change `permit_params :title` to `aaapermit_param :title`.
You get `undefined method 'aaapermit_params' for an instance of ActiveAdmin::ResourceDSL` as expected.
Now change `aaapermit_params` back to `permit_params` and reload the page.

### Expected behavior

The page should reload.

### Actual behavior

`superclass mismatch for class DashboardController`

A quick patch that I discovered is to add this code to the gem:

```diff
# lib/active_admin/namespace.rb

    # TODO: replace `eval` with `Class.new`
    def register_resource_controller(config)
+     name_parts = config.controller_name.split('::')
+     const_name = name_parts.last
+     if Admin.const_defined?(const_name)
+       Admin.send(:remove_const, const_name)
+     end
      eval "class ::#{config.controller_name} < ActiveAdmin::ResourceController; end"
      config.controller.active_admin_config = config
    end
```

**Comment by giuseb** [comment]
I have observed the same behavior

**Comment by aalyahya** [comment]
I've experienced it as well

**Comment by vfonic** [comment]
I've been seeing this happen many times in development. I believe it happens when there's a syntax error in the code (for example an extra `end`) so the code doesn't get reloaded correctly. This makes the code enter in "bad state". Even after fixing the syntax error, the issue persists. The admin doesn't load with `superclass mismatch for class XXX`. The only way to fix the issue is to restart the Rails server.

`XXX` is the alphabetically first resource. It could be `DashboardController` or it could be `ArticleController` if you have `Article` resource in your app.

@javierjulio I believe this is a bug and should be an issue rather than a discussion.

**Comment by sbounmy** [comment]
Having the same issue here too
